Refactor + CLI restructure

robertgzr opened the pull request:
The README is up to date (at least the quickstart section) so that can be used for testing as the way we interact with kube-spawn has changed quite a lot:

- `kube-spawn create`: used to generate all of the scripts and configs we need in the machines.
Everything is put into `/var/lib/kube-spawn/default` (tweakable with the `--dir` and `--cluster-name` flags).
It also outputs a `kspawn.toml` file into the root of that directory that contains the parameters used by kube-spawn (node number, base image, k8s version, etc.)
Most important things can be changed on create using flags and environment variables.

- `kube-spawn start`: takes the generated files and starts the machines and runs `kubeadm` on them. After this you should have a working cluster

- `kube-spawn stop`: brings down everything (does not retain the cluster as we delete the base images for the nodes)

- `kube-spawn destroy`: deletes `/var/lib/kube-spawn/default`

robertgzr commented:
CI fails because I don't have my [fork of `go-toml`](https://github.com/robertgzr/go-toml) checked into vendor/

dongsupark commented on pkg/config/types.go:
```diff
@@ -0,0 +1,98 @@
+	// Files to be copied from host to overlay of all machines
+	// map: dst = src
+	// not saved into kspawn.toml
+	Copymap map[string]string `toml:"-"`
```
Is there any reason to do it like `dst = src`? Because to me it looks intuitive to have the other way around, `src = dst`.

robertgzr replied:
Yeah I thought about both ways... I ended up going with this because it feels like variable assignment to me. 
But I am open to changing it if that makes no sense.

dongsupark replied:
@robertgzr Ok, if you say so, I don't have strong opinion.

robertgzr commented:
I tried splitting up the big commit into logical chunks, commited the vendor fixes, added a quick little list subcommand and fixed the panics related to running with `rkt` as container runtime

robertgzr commented:
Things that remain todo:
- [ ] clean up cni config (probably in a later PR)
- [ ] terminal output broken after `kube-spawn start`

robertgzr commented:
Rebased and replaced the delay in the stop commad with a retry loop

schu commented on pkg/config/defaults.go:
```diff
@@ -0,0 +1,257 @@
+	DefaultRuntimeTimeout        = "15m"
+	DefaultRktStage1ImagePath    = "/usr/lib/rkt/stage1-images/stage1-coreos.aci"
+
+	CacheDir = ".cache"
```
I don't think the directory should be hidden, as it leaves in `/var/lib/kube-spawn`.

robertgzr replied:
I wanted some kind of prefix as this makes distinguishing and [listing](https://github.com/kinvolk/kube-spawn/pull/161/commits/5b437dab9961f6c734701fd05c521aeecc0860a4#diff-04a5247871266df765044434b04e51acR58) cluster envs easier.

schu replied:
The better way would be then to have an `environments` subfolder. Can be changed in the next version.

schu commented on pkg/config/defaults.go:
```diff
@@ -0,0 +1,257 @@
+	// NOTE: K8s 1.8 or newer fails to run by default when swap is enabled.
+	// So we should disable the feature with an option "--fail-swap-on=false".
+	if !cfg.DevCluster && utils.CheckVersionConstraint(cfg.KubernetesVersion, "<1.8.0") {
+		cfg.RuntimeConfiguration.FailSwapOn = true
```
I have asked this before: can we not disable swap for the machines instead?

dongsupark replied:
As for the swap, I think it's hard to expect every user to disable swap. It might be easier to do that if we are only dealing with servers, or if most Linux distros disable swap out of the box. Sadly both are not the case.
So it's kind of nasty, but I would prefer to get it handled in this way on the k8s side.

schu replied:
I was thinking of using systemd properties (`systemd-nspawn --property ...`) to disable it for the containers, not for the host system :) On a cursory look, I haven't found a way to do this from the CLI though. So I'm not sure it's possible. IIRC LXD does provide an option for that.

robertgzr commented on cmd/kube-spawn/create.go:
```diff
@@ -0,0 +1,165 @@
+	createCmd.Flags().StringP("image", "i", "", "base image for nodes")
+	viper.BindPFlags(createCmd.Flags())
+
+	viper.BindEnv("runtime-config.rkt.rkt-bin", "KUBE_SPAWN_RKT_BIN")
```
maybe this is just me but I don't like the idea of having cli flags for all of these binaries. It's more or less necessary for the stage1 image but if you have installed rkt with your distros pkg manager and/or have `$GOPATH/bin` in your `PATH` you won't need to configure it anyway...

robertgzr replied:
Also we have the problem that not all of the options in the config file are possible to be set from env/cli (or make sense to be) - like the path mapping for example.

This might be solvable by having two structures, one that viper binds to and the other we handle ourselves.

schu replied:
Yeah, probably user and cluster configuration should be split.

robertgzr commented on cmd/kube-spawn/start.go:
```diff
@@ -0,0 +1,161 @@
+	log.Printf("spawning cluster %q (%d machines)", cfg.Name, cfg.Nodes)
+
+	// TODO: find a place for these calls
+	tmp(cfg)
```
will fix this

robertgzr commented on cmd/kube-spawn/stop.go:
```diff
@@ -18,88 +18,117 @@ package main
+func removeImage(machineName string) error {
+	done := false
+	retries := 0
+	for !done {
```
Looks like this also applies to https://github.com/kinvolk/kube-spawn/blob/master/pkg/nspawntool/run.go#L106

iaguis commented on pkg/config/defaults.go:
```diff
@@ -0,0 +1,257 @@
+	pms := []Pathmap{
+		{Dst: "/usr/bin/rkt", Src: cfg.RuntimeConfiguration.Rkt.RktBin},
+		{Dst: "/usr/bin/rktlet", Src: cfg.RuntimeConfiguration.Rkt.RktletBin},
+		{Dst: "/usr/bin/stage1-coreos.aci", Src: cfg.RuntimeConfiguration.Rkt.Stage1Image},
```
can we change this so it works if I use a non-coreos stage1? (see https://github.com/kinvolk/kube-spawn/pull/162)

robertgzr replied:
forgot about this, thanks

robertgzr commented:
Ready for merge 🎉

blixtra commented:
I notice the `up` command was removed. Let's please add that back. We should have a one-command way to get the cluster from 0 to up.

We can add that back in another PR.
